### ISSUE TITLE
Research: Duplex mode on 24GB GPU — runtime system-prompt switching, n_ctx sub-model limit, KV pruning continuation, VRAM requirement

### ISSUE THREAD

**Issue by ygwfdjts6k-dotcom** [tc-mb/llama.cpp-omni#88]
### Research Stage

- [ ] Background Research (Let's try to avoid reinventing the wheel)
- [ ] Hypothesis Formed (How do you think this will work and it's effect?)
- [ ] Strategy / Implementation Forming
- [ ] Analysis of results
- [ ] Debrief / Documentation (So people in the future can learn from us)

## Environment
- Model: MiniCPM-o-4_5-Q4_K_M.gguf (Q4_K_M)
- GPU: RTX 3090 24GB, CUDA 13.1, gcc 15
- Build: repo snapshot ~2026-07-07, duplex_mode=1, -c 4096 -ngl 99
- Use case: real-time telephony (FreeSWITCH), one persistent duplex session per call

The official `llama-omni-test-duplex` binary runs correctly on this setup (multi-turn dialogue, exit 0). My questions are about design/capability, not a crash.

## Q1 — Runtime system-prompt switching (biggest blocker)
`omni_voice_clone_prompt` / `omni_assistant_prompt` are hardcoded in `omni.cpp` (~line 4064-4069) and prefilled during `omni_duplex_session_begin`. My app serves multiple tenants, each needing a different persona/system prompt.

Is there a supported way to set the system prompt at session-begin time (per session) rather than at compile time? If not, is re-running `session_begin` with a modified `voice_clone_prompt` the recommended approach, and what's the cost/correctness impact on the prewarmed KV?

For reference, my bridge currently passes only the following at session begin:

    {"cmd":"begin"}

If `session_begin` already accepts a system-prompt / persona field I'm not using, please point me to it.

## Q2 — n_ctx vs sub-model n_ctx_train=4096
Startup logs show two different `n_ctx_train`: main LLM = 40960, but another sub-model = 4096.

With `-c 8192`, speak-count on the same audio is unstable across runs (3/5/11/9 over 4 runs); with `-c 4096` it's stable (8/8/8 over 4 runs). "speak-count" = number of speak events counted from stderr/stdout over the same fixed test set.

Is duplex output hard-limited by the 4096 sub-model, i.e. is `-c 4096` the correct ceiling? Can the main LLM use a larger context independently?

## Q3 — KV pruning continuation
README says to stop on KV pruning. In telephony, pruning triggers at n_past≈2000 (~2.5 min). Is there a recommended way to continue a session past pruning without ending the call?

## Q4 — VRAM requirement
The HF model card states duplex needs 28GB; my 3090 has 24GB and runs without OOM. Is 28GB a hard requirement, or is 24GB acceptable with Q4_K_M? Any known instability from running under the stated requirement?

## Reproduce (Q2)
Run each 4x, count speak events per run:

    ./build/bin/llama-omni-test-duplex \
      -m MiniCPM-o-4_5-Q4_K_M.gguf --ref-audio ref.wav \
      -c 4096 --test /path/chunk_ 100

    ./build/bin/llama-omni-test-duplex \
      -m MiniCPM-o-4_5-Q4_K_M.gguf --ref-audio ref.wav \
      -c 8192 --test /path/chunk_ 100

### Relevant log output

```shell

```

**Comment by tc-mb** [comment]
Hi, thanks for the detailed write-up — answering your four questions below.

**Q1 — Runtime system-prompt switching**
Yes. Multi-session setups can use different system prompts; this is supported and intended to be configurable on your side (e.g. per-session `voice_clone_prompt` / `assistant_prompt` on the server path). You can tune persona / system prompt per tenant or per call as needed.

**Q2 — `n_ctx` vs the 4096 sub-model**
You do not have to lock to 4096. Other lengths such as 8k are fine.

**Q3 — Continuing after KV pruning**
You can keep using the session after KV pruning. The trade-off is that earlier conversation memory carried in the pruned KV will be lost; the call does not need to end, but prior context may no longer be available.

**Q4 — VRAM / 28GB**
The “at least 28GB” note on the HF / PyTorch docs is for the **PyTorch** inference path, not for the llama.cpp-omni GGUF path.
For **C++ + Q4_K_M**, duplex is OK around **~11GB VRAM**. We ship Q4_K_M so more consumer GPUs can run it; we’ve also checked quality (including human evaluation) for this setting. So a 24GB 3090 with Q4_K_M is expected to work — 28GB is not a hard requirement for the quantized C++ path.

Feel free to follow up if anything is still unclear.

**Comment by ygwfdjts6k-dotcom** [comment]
Thanks for the detailed answers — Q1 and Q4 are fully resolved on my side. Two follow-ups.

## Q2 follow-up — "supported" vs "stable"

My original question was ambiguous, so let me restate it. I'm not asking whether 8192 is *allowed*; I'm asking about the run-to-run variance I observed at 8192.

Same binary, same model file, same `ref.wav`, same fixed test set, 4 runs each, counting speak events from stderr/stdout:

| `-c` | run 1 | run 2 | run 3 | run 4 |
|------|-------|-------|-------|-------|
| 4096 | 8 | 8 | 8 | 8 |
| 8192 | 3 | 5 | 11 | 9 |

With identical input and no sampling changes between runs, I would expect the speak-event count to be deterministic, or at least tightly clustered. At 4096 it is; at 8192 it varies by roughly 4x.

Questions:

1. Is this variance expected at 8192, or does it indicate something misconfigured on my side?
2. The startup logs show `n_ctx_train = 40960` for the main LLM but `4096` for the other sub-model. Does the duplex path route anything through the 4096 sub-model in a way that would make contexts beyond 4096 behave differently?
3. Is there a recommended way to verify that a given `-c` value is safe for the duplex path, beyond "it doesn't crash"?

This matters for my use case (real-time telephony) because an unstable speak-event count means the assistant sometimes doesn't speak during a live call. I'm currently pinning `-c 4096` based on this measurement, and I'd rather understand the cause than keep the pin as a superstition.

## Q3 follow-up — updating the system prompt mid-session

Your Q1 answer says `voice_clone_prompt` / `assistant_prompt` are configurable per session. Your Q3 answer says the call can continue past KV pruning, but earlier conversation memory carried in the pruned KV is lost.

Combining the two: is there a supported way to **update** the system prompt on an already-running session, rather than only at `session_begin`?

The pattern I have in mind: as pruning approaches, summarize the conversation so far and re-inject that summary as updated system-prompt context, so the assistant retains key facts (budget, timeline, commitments) that would otherwise be dropped.

If mid-session update isn't supported, is there another recommended mechanism for carrying salient context across a pruning boundary — or is ending and re-beginning the session with an updated prompt the intended approach? If the latter, roughly what does re-running `session_begin` cost on an already-prewarmed process?

**Comment by tc-mb** [comment]
Thanks for the follow-up.

Q2: We could not reproduce an 8192-specific issue. With 100 chunks, both 4096 and 8192 produced identical results across four runs. In longer tests, both varied with temperature 0.8, while temperature 0 was fully deterministic. This looks like sampling variance amplified over long sessions, not a 4096 sub-model limit. The 4096 value belongs to the TTS model; SPEAK/LISTEN is decided by the main LLM.

Q3: Updating the system prompt during an active session is not supported. Restarting with a new prompt creates a new session and loses the existing KV context. This is a limitation of the current model generation and will be improved in the next Omni generation.

**Comment by ygwfdjts6k-dotcom** [comment]
# Follow-up: audio-only duplex for telephony — TTS tail misalignment, system prompt, SPEAK/LISTEN tuning, and whether this path is viable

Thanks for the earlier answers — Q1 (per-session prompt) and Q4 (VRAM) are resolved, and the `temperature 0.8` explanation for run-to-run variance was very helpful.

I've since put duplex into a real telephony deployment. New findings and questions below. The last section (Q5) is the one that matters most to me.

---

## Setup

- Model: `MiniCPM-o-4_5-Q4_K_M.gguf` · GPU: single RTX 3090 24GB
- Build: repo snapshot 2026-07-21, `master` branch, `duplex_mode=1`, `-c 4096`, `-ngl 99`
- Custom C++ host built directly on the `omni` API (not `llama-omni-server`): one process per session, stdin/stdout protocol, 2 concurrent sessions per GPU
- **`media_type = 1` (audio only — this is a phone call, there is never any video)**
- Audio: 8kHz telephony → FIR-upsampled to 16kHz → `omni_duplex_push_frame`, 1s chunks
- TTS wavs are picked up by the telephony layer and streamed to the caller as PCM
- VRAM: 9.3GB per session, 18.7GB for two — consistent with your ~11GB figure

`llama-omni-test-duplex` runs correctly on this machine (exit 0, multi-turn). Decode latency matches your published benchmarks (3 tokens ≈ 115ms measured vs ~115ms documented on 4090/F16).

---

## Q1 — A single sentence produces N+1 wavs, and the tail chunk arrives too late to concatenate

One spoken sentence produced **three** wav files, but the LLM emitted only **two** text chunks.

`llm_debug/llm_text.txt`:

    [chunk_0] 你好，有什么
    [chunk_1] 我可以帮你的吗？

`llm_debug/llm_token_ids.txt`:

    [chunk_0] 108386 3837 104139
    [chunk_1] 109944 99663 103929 101037 11319

Wav files written (mtime, duration, format):

    09:46:08.142  wav_4000.wav  0.84s  24kHz mono  -> chunk_0
    09:46:09.221  wav_5001.wav  1.00s  24kHz mono  -> chunk_1
    09:46:09.914  wav_6002.wav  0.76s  24kHz mono  -> no corresponding text chunk

`wav_6002.wav` has no matching entry in `llm_text.txt`. I assume it's the Token2Wav sliding-window tail (28 tokens in / 25 stride).

The problem: it lands **276ms after** playback of the sentence has already started, so it can't be merged with the first two. It gets played as a separate buffer, and the extra `broadcast` boundary is audible as a gap **in the middle of one sentence**.

**Questions:**

1. Is there a signal that means "this is the final wav for this turn"? I can see `current_turn_ended`, `ended_with_listen` and `current_turn_id` in `omni.h` — is one of these the intended end-of-turn indicator on the C++ API, and is it safe to read from a consumer thread while decode is running?
2. Is there a supported way to make Token2Wav flush a whole semantic unit at once, so one sentence arrives as one wav instead of N+1 pieces?
3. Telephony is a fixed-rate continuous stream, not a file player. The official demos watch a directory and play files as they appear, which assumes a forgiving playback layer. What buffering strategy do you recommend for a fixed-rate channel?

---

## Q2 — Is `media_type = 1` (audio-only) a supported duplex configuration?

All duplex documentation describes the proactive mechanism as the LLM monitoring **video and audio** streams at 1Hz, and the HTTP API guide says each prefill iteration should carry one audio chunk **and one image frame**.

My deployment has no video, ever.

SPEAK rate measured over three real calls (SPEAK events / total decode chunks):

| call | speak | total | rate |
|---|---|---|---|
| 1 | 12 | 118 | 10.2% |
| 2 | 25 | 396 | 6.3% |
| 3 | 11 | 155 | 7.1% |

With `llama-omni-test-duplex` on a fixed 100-chunk test set, SPEAK count was 6/100 vs 10/100 depending **only** on which `--ref-audio` file I used (ref audio length also shifted `n_past` from 133 to 268, so the system-prompt budget changes with it).

Replies are also very short and generic — "好的。" / "可以啊。" / "当然可以的。" — and rarely engage with what the caller actually said.

**Questions:**

1. Is audio-only duplex (`media_type=1`) a supported and tested configuration, or is the proactive-interaction mechanism trained specifically for vision+audio?
2. If video genuinely matters to the decision layer, is feeding a static dummy frame better than sending no video at all?
3. Is a single-digit SPEAK rate expected for audio-only duplex, or does it indicate a misconfiguration on my side?

---

## Q3 — The duplex system prompt is hardcoded to English

In `omni.cpp` (~4064) the duplex branch sets:

```cpp
audio_voice_clone_prompt = "<|im_start|>system\nStreaming Duplex Conversation! You are a helpful assistant.\n<|audio_start|>";
audio_assistant_prompt   = "<|audio_end|><|im_end|>\n";
```

while the non-duplex branch gets a much richer Chinese prompt (模仿音频样本的音色… 请认真、高质量地回复用户的问题…).

`omni_set_language()` makes it explicit:

```cpp
// 双工模式：prompt 固定使用英文（与 Python 对齐）
```

Your earlier answer said per-session `voice_clone_prompt` / `assistant_prompt` is supported and intended to be configurable. For the duplex path these two seem to point in different directions.

**Questions:**

1. Is the English-only duplex prompt a deliberate constraint (e.g. SPEAK/LISTEN was trained with that exact prefix), or just a default?
2. If I overwrite these fields on `ctx_omni` after `omni_init` and before `omni_duplex_session_begin`, is that supported? Any invariant I must preserve — must it start with `<|im_start|>system`, must the marker "Streaming Duplex Conversation!" be present, must it remain English?
3. For a Chinese-language phone assistant, would a Chinese duplex system prompt be expected to degrade SPEAK/LISTEN behaviour?

---

## Q4 — Tuning turn-taking: temperature, `listen_prob_scale`, `force_listen_count`

Following your note that variance came from `temperature 0.8` and that SPEAK/LISTEN is decided by the main LLM: since `<|speak|>` / `<|listen|>` are sampled tokens, temperature should affect the gating decision too, not only the wording.

I also found in `omni.h` / `omni.cpp`:

- `listen_prob_scale` (default `1.0f`), applied as `listen_bias = (scale - 1.0f) * 2.0f` added to the `<|listen|>` logit
- `force_listen_count` (default `3`) — this matches my logs exactly; the model never speaks before the 4th chunk

**Questions:**

1. For a live phone conversation, what `temperature` do you recommend? Does lower temperature make the assistant speak more often, or less?
2. Is `listen_prob_scale < 1.0` (biasing *against* `<|listen|>`, i.e. toward speaking) a supported use of the parameter, or was it only intended for values `> 1.0`? Is there a safe range?
3. Are these parameters sufficient to tune turn-taking, or is gating behaviour essentially fixed for this model generation?

---

## Q5 — Is this path viable for a production phone assistant, and what is the optimal setup?

I'll be direct about my position. I'm building a commercial outbound-calling assistant. The benchmark I'm judged against is the cascaded systems shipped by larger vendors — dedicated streaming ASR + a large LLM + commercial TTS, plus substantial dialogue-state engineering. Those feel close to human.

What duplex currently gives me is far from that: single-digit SPEAK rate, replies collapsing into short acknowledgements, and audible gaps inside a single sentence.

I'd rather make an informed engineering decision than keep tuning blindly.

1. Is end-to-end duplex in this generation intended to reach product-grade conversational quality, or is it best understood as a research preview of the architecture — with the cascaded pipeline still the right choice for shipping voice products today?

2. If duplex *can* reach product quality for audio-only phone calls, what configuration would you actually recommend? Concretely: `media_type`, `n_ctx`, `temperature`, `listen_prob_scale`, `force_listen_count`, system-prompt language and content, ref-audio length, and whether `master` or `feat/web-demo` is the right branch for this use case. (Issue #74 suggests the two branches are incompatible duplex engines — I'm on `master`.)

3. Which of my symptoms are known limitations of this generation, and which are likely misconfiguration on my side?
   - SPEAK rate of 6–10% on audio-only input
   - replies degenerating into short acknowledgements rather than substantive answers
   - one sentence arriving as N+1 wavs with a late tail chunk

4. Is fine-tuning on domain dialogue data the intended way to close this gap? If so, is there guidance for fine-tuning the duplex / proactive-interaction behaviour specifically, as opposed to base conversational ability?

5. Is there any reference deployment of MiniCPM-o duplex in a telephony or customer-service setting whose numbers I could compare against?

---

I'm happy to share full logs, the wav files, or a minimal reproduction if that helps. Thanks for your time — the previous answers were genuinely useful.

**Comment by tc-mb** [comment]
Thanks for the detailed feedback.

**Q1:** Duplex TTS is designed for streaming playback. Audio chunks should be consumed and played as they are generated, rather than waiting for the complete sentence and concatenating all WAV files first. Producing multiple audio chunks, including a final tail chunk, is expected. For telephony, the playback side should feed decoded PCM into a continuous audio queue or jitter buffer instead of treating each WAV as a separate playback operation.

**Q2:** Audio-only duplex is supported. Please first try the audio duplex mode in our official demo and compare its behavior with your custom C++ host. This should help determine whether the low SPEAK rate comes from the model or from differences in the integration.

**Q3:** The duplex system prompt can be customized. We recommend preserving the special-token structure and prompt format, while modifying the natural-language content between those markers. You can test different Chinese or domain-specific prompt contents for your use case.

**Q4:** We have not specifically tested these parameters in a telephony deployment. The appropriate values may depend heavily on the audio segmentation, turn-taking behavior, and business scenario, so you will likely need to benchmark different parameter combinations on representative call data.

**Q5:** MiniCPM-o is an open-source model. If you require production-grade commercial deployment, you can contact us to discuss commercial development and optimization. The current open-source version, including its approximately 8K context capability, may not be sufficient by itself for a production telephone assistant, especially for long calls and complex dialogue-state requirements.
